### ISSUE TITLE
DelegateDecompiler.EntityFramework with Odata v4

### ISSUE THREAD

**Issue by eabrahamyan** [hazzik/DelegateDecompiler#51]
Hi I have a project where I use EF and I  have some computed properties. It works fine until I use Odata $top or $skip query options.

``` json
{
  "error":{
    "code":"","message":"An error has occurred.","innererror":{
      "message":"Object of type 'DelegateDecompiler.DecompiledQueryable' cannot be converted to type 'System.Linq.IQueryable`1[Entity]'.","type":"System.ArgumentException","stacktrace":"   at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)\r\n   at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)\r\n   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)\r\n   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)\r\n   at System.Web.OData.Query.Expressions.SelectExpandBinder.Bind(IQueryable queryable)\r\n   at System.Web.OData.Query.ODataQueryOptions.ApplyTo(IQueryable query, ODataQuerySettings querySettings)\r\n   at System.Web.OData.EnableQueryAttribute.ExecuteQuery(Object response, HttpRequestMessage request, HttpActionDescriptor actionDescriptor)\r\n   at System.Web.OData.EnableQueryAttribute.OnActionExecuted(HttpActionExecutedContext actionExecutedContext)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.OnActionExecutedAsync(HttpActionExecutedContext actionExecutedContext, CancellationToken cancellationToken)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()"
    }
  }
}
```

i'll appreciate any help.
Thank you.

**Comment by DreadedTuba** [comment]
Hi, How you are using DelegateDecompiler with OData?  I'm having difficulties integrating them.  I'm using the Asp.Net Odata framework.  I almost got it working but ran into a problem with JSON serialization when $top and $filter options are used together (works separately, but not together).  The serializer complains about a nil value.  I found a related article that said it was related to IQueryable and IQueryable<T> conversion.  The symptom is similar to this issue, but my error message it different.
Thanks.

**Comment by grassgit** [comment]
I ran in to this issue and incase someone else runs into it to I did figure out a workaround.

**TL;DR** the solution is a replacement of the EnableQuery below (with some restrictions)

With some decompiling through the WebAPI OData code this issue is actually cause by the appliction of the odata select, expand and (model) pagesize options. During these the an up-cast is done which works for the standard EntityFramework but not if the queryable is the result of an AutoMapper ProjectToQueryable<T> call. This is because of the use of the DelegateDecompiler, but this is required because of the way the tree walk of IQueryable exectution works. The only way to avoid this would be the execute the IQueryable before the select, expand or pagesize options are applied. This obviously negates the advantage of having these as an IQueryable as the entire purpose of the IQueryable with the automapper is to apply the query on your model at the database so only the required data is retrieved from the data store.

The workaround below replaced the EnableQuery attribute. I have not tested all options and scenarios with this solution and there is a major restriction to the odata options.

**Restrictions:**: 
This workaround work by instructing the OData query application to ignore any $select and $expand filters and the PageSize as defined on the model. 
For proper use you EDM model builder should also disable the options to this way the users get an error that the option is disabled rather than returning unexpected results.

For my application this approach is acceptable as I would prefer applying equivalents of the the select and expand options in the model. PageSize can be replaced by use of the top and skip option.

**Workaround**
Add the following to your project:
```
public class EnableQueryAutoMappedAttribute : EnableQueryAttribute
{

    public override IQueryable ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
    {
        var options = new ODataQuerySettings();
        options.EnableConstantParameterization = EnableConstantParameterization;
        options.EnsureStableOrdering = EnsureStableOrdering;
        options.HandleNullPropagation = HandleNullPropagation;
        // Ignore pagesize (options.PageSize )

        // Ignore select or expand. These cause the follow error:
        // ArgumentException: Object of type 'DelegateDecompiler.DecompiledQueryable' cannot be converted to type 'System.Linq.IQueryable`1[Entity] 
        var query = queryOptions.ApplyTo(queryable, options, AllowedQueryOptions.Select | AllowedQueryOptions.Expand);
        return query;

        // Call IQueryProvider.CreateQuery. If this is not called this causes an error:
        // InvalidOperationException: (The 'ObjectContent`1' type failed to serialize the response body for content type 'application/json; odata.metadata=minimal')
        // The CreateQuery method is generic and must be called with the appropriate type parameter (object will not work).
        // This requires reflection as attributes cannot be generic.
        var createQuery = typeof(IQueryProvider).GetMethods().Single(x => x.Name == "CreateQuery" && x.IsGenericMethodDefinition);
        createQuery = createQuery.MakeGenericMethod(query.ElementType);
        return (IQueryable)createQuery.Invoke(query.Provider, new[] { query.Expression });
    }
}
```
On the ODataController or methods replace the EnableQuery with EnableQueryAutoMapped. e.g.:
```
[EnableQueryAutoMapped]
public IQueryable<MyType> Get()

```

**Comment by hazzik** [comment]
Shall be fixed with #97
